lib/fileset/README.md

```diff
@@ -41,9 +41,14 @@ An attribute set with these values:
 - `_type` (constant string `"fileset"`):
   Tag to indicate this value is a file set.
 
-- `_internalVersion` (constant `2`, the current version):
+- `_internalVersion` (constant `3`, the current version):
   Version of the representation.
 
+- `_internalIsEmptySingleton` (bool):
+  Whether this file set is the singleton representing an empty file set.
```

Comment:
```suggestion
- `_internalIsEmpty` (bool):
  Whether this file set is the empty file set.
```

Not sure if it's required to call it a singleton in the parameter. It's just *the* root-less empty set, regardless of how it's implemented.

Reply:
I wanted to avoid calling it that because there's multiple ways to represent an empty fileset: `null` as the tree is another representation, `{ }` is another and there's many more if you have some files in the base path.

Reply:
What about this then?

```suggestion
- `_internalIsEmptyRootless` (bool):
  Whether this file set is the empty file set without a base path.
```

Reply:
I like that, should be `_intetrnalIsEmptyBaseless` though. Will change

Reply:
Although, it's not _just_ without a base path, it's also without a file set tree..

Reply:
Is it even possible to have a file set with a tree but without a base?

Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

Reply:
At least currently no base path also means no tree and the other way around, so yeah it might be okay to word it like that. In the end it doesn't matter much since indeed it is internal.

> Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

It's not relevant for the user, it's an implementation detail. To the user it behaves just like they would expect. Exposing this as a function would expose implementation details.

Reply:
To be clear, that helper function would also be internal.

Reply:
Besides being returned for an empty union, isn't its unique property that its the identity of any union operation?
You may consider `_isUnionIdentity`.